test/mode_permutation_test.py

```diff
@@ -0,0 +1,170 @@
+__author__ = 'Tomasz Rybotycki'
+
+# TR TODO: Consider making this an actual test.
+
+import numpy as np
+import math
+from scipy import special
+from src.Boson_Sampling_Utilities import modes_state_to_particle_state, \
+    particle_state_to_modes_state, calculate_permanent, generate_lossy_inputs
+from src.BosonSamplingSimulator import BosonSamplingSimulator
+from src.simulation_strategies.FixedLossSimulationStrategy import FixedLossSimulationStrategy
+
+# Generate permutation matrix and define initial state.
+permutation_matrix = np.zeros((5, 5))
+initial_state = [1, 1, 1, 0, 0]
+
+permutation_matrix[0][2] = 1
+permutation_matrix[1][0] = 1
+permutation_matrix[2][3] = 1
+permutation_matrix[3][4] = 1
+permutation_matrix[4][1] = 1
+
+# Define some constants that will be used through the experiment.
+NUMBER_OF_LOST_PARTICLES = 2
+INITIAL_NUMBER_OF_PARTICLES = initial_state.count(1)
+NUMBER_OF_PARTICLES_LEFT = INITIAL_NUMBER_OF_PARTICLES - NUMBER_OF_LOST_PARTICLES
+NUMBER_OF_MODES = 5
+
+# Define some variables for this scenario, to ease the computations.
+possible_outcomes = [[1, 0, 0, 0, 0], [0, 1, 0, 0, 0], [0, 0, 0, 0, 1]]
+
+
+def main():
+    exact_distribution = calculate_exact_distribution()
+    print(f'Exact distribution: {exact_distribution}')
+
+    approximate_distribution = calculate_approximate_distribution()
+    print(f'Approximate distribution: {approximate_distribution}')
+
+
+def calculate_exact_distribution() -> list:
+    """
+    This method will be used to calculate the exact distribution of lossy boson sampling experiment.
+    The results will be returned as a table of probabilities of obtaining the outcome at i-th index.
+    :return: List of probabilities of outcomes.
+    """
+    outcomes_probabilities = []
+    for outcome in possible_outcomes:
+        outcomes_probabilities.append(calculate_probability_of_outcome(outcome))
+    return outcomes_probabilities
+
+
+def calculate_probability_of_outcome(outcome: list) -> float:
+    """
+    Given input state and losses as described in the constants of this file calculate the probability
+    of obtaining the outcome given as a parameter, after traversing lossy linear-optical channel.
+    :param outcome: An outcome which probability of obtaining will be calculated.
+    :return: Probability of obtaining given outcome in situation presented by by the
+    """
+    outcome_probability = 0  # Initialize with 0 for later debug purposes.
+    outcome_state_in_particle_basis = modes_state_to_particle_state(outcome, NUMBER_OF_PARTICLES_LEFT)
+
+    outcome_probability = calculate_probability_of_outcome_state_for_indistinguishable_photons(outcome_state_in_particle_basis)
+
+    # Different states in particles-basis may give the same outcome state.
+    outcome_probability *= math.factorial(NUMBER_OF_PARTICLES_LEFT)
```

Comment:
No I did not. Should I? I believe that nothing wrong can be expected here, but maybe I'm wrong...

Reply:
The only thing is that the factorial gets pretty slow with the number of iterations and size of the value that you count factorial for. With caching, these values are not calculated, just fetched from memory if they were calculated once. If we're in thousands of iterations and tens of values that have factorial calculated for them, it's nothing, but once we move by two orders of magnitude, this could mean some serious performance boost.

Reply:
Oh, I see. In that case, there's no need for caching here, but I'll keep that in mind in case I write some code that would calculate more and larger factorials. Thanks!